### ISSUE TITLE
Unable to create failover group for SQL Managed Instance

### ISSUE THREAD

**Issue by fcatacut** [hashicorp/terraform-provider-azurerm#15556]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform (and AzureRM Provider) Version

```
Terraform v1.0.5
+ provider registry.terraform.io/hashicorp/azurerm v2.97.0
```

### Affected Resource(s)

* `azurerm_mssql_managed_instance_failover_group`
* `azurerm_mssql_managed_instance`

### Terraform Configuration Files

```hcl
provider "azurerm" {
  features {}
}

variable "instances" {
  default = {
    primary = {
      location       = "eastus2"
      address_space  = "10.1.0.0/16"
      address_prefix = "10.1.0.0/24"
    }
    secondary = {
      location       = "centralus"
      address_space  = "10.2.0.0/16"
      address_prefix = "10.2.0.0/24"
    }
  }
}

variable "tags" {
  default = {}
}

resource "azurerm_resource_group" "this" {
  for_each = var.instances
  name     = format("example-%s-rg", each.key)
  location = each.value.location
  tags     = var.tags
}

resource "azurerm_network_security_group" "this" {
  for_each            = var.instances
  name                = format("example-%s-nsg", each.key)
  resource_group_name = azurerm_resource_group.this[each.key].name
  location            = each.value.location
  tags                = var.tags
}

resource "azurerm_network_security_rule" "allow_management_inbound" {
  for_each                    = var.instances
  name                        = "allow_management_inbound"
  priority                    = 106
  direction                   = "Inbound"
  access                      = "Allow"
  protocol                    = "Tcp"
  source_port_range           = "*"
  destination_port_ranges     = ["9000", "9003", "1438", "1440", "1452"]
  source_address_prefix       = "*"
  destination_address_prefix  = "*"
  resource_group_name         = azurerm_resource_group.this[each.key].name
  network_security_group_name = azurerm_network_security_group.this[each.key].name
}

resource "azurerm_network_security_rule" "allow_misubnet_inbound" {
  for_each                    = var.instances
  name                        = "allow_misubnet_inbound"
  priority                    = 200
  direction                   = "Inbound"
  access                      = "Allow"
  protocol                    = "*"
  source_port_range           = "*"
  destination_port_range      = "*"
  source_address_prefix       = "10.0.0.0/24"
  destination_address_prefix  = "*"
  resource_group_name         = azurerm_resource_group.this[each.key].name
  network_security_group_name = azurerm_network_security_group.this[each.key].name
}

resource "azurerm_network_security_rule" "allow_health_probe_inbound" {
  for_each                    = var.instances
  name                        = "allow_health_probe_inbound"
  priority                    = 300
  direction                   = "Inbound"
  access                      = "Allow"
  protocol                    = "*"
  source_port_range           = "*"
  destination_port_range      = "*"
  source_address_prefix       = "AzureLoadBalancer"
  destination_address_prefix  = "*"
  resource_group_name         = azurerm_resource_group.this[each.key].name
  network_security_group_name = azurerm_network_security_group.this[each.key].name
}

resource "azurerm_network_security_rule" "allow_tds_inbound" {
  for_each                    = var.instances
  name                        = "allow_tds_inbound"
  priority                    = 1000
  direction                   = "Inbound"
  access                      = "Allow"
  protocol                    = "Tcp"
  source_port_range           = "*"
  destination_port_range      = "1433"
  source_address_prefix       = "VirtualNetwork"
  destination_address_prefix  = "*"
  resource_group_name         = azurerm_resource_group.this[each.key].name
  network_security_group_name = azurerm_network_security_group.this[each.key].name
}

resource "azurerm_network_security_rule" "deny_all_inbound" {
  for_each                    = var.instances
  name                        = "deny_all_inbound"
  priority                    = 4096
  direction                   = "Inbound"
  access                      = "Deny"
  protocol                    = "*"
  source_port_range           = "*"
  destination_port_range      = "*"
  source_address_prefix       = "*"
  destination_address_prefix  = "*"
  resource_group_name         = azurerm_resource_group.this[each.key].name
  network_security_group_name = azurerm_network_security_group.this[each.key].name
}

resource "azurerm_network_security_rule" "allow_management_outbound" {
  for_each                    = var.instances
  name                        = "allow_management_outbound"
  priority                    = 102
  direction                   = "Outbound"
  access                      = "Allow"
  protocol                    = "Tcp"
  source_port_range           = "*"
  destination_port_ranges     = ["80", "443", "12000"]
  source_address_prefix       = "*"
  destination_address_prefix  = "*"
  resource_group_name         = azurerm_resource_group.this[each.key].name
  network_security_group_name = azurerm_network_security_group.this[each.key].name
}

resource "azurerm_network_security_rule" "allow_misubnet_outbound" {
  for_each                    = var.instances
  name                        = "allow_misubnet_outbound"
  priority                    = 200
  direction                   = "Outbound"
  access                      = "Allow"
  protocol                    = "*"
  source_port_range           = "*"
  destination_port_range      = "*"
  source_address_prefix       = "10.0.0.0/24"
  destination_address_prefix  = "*"
  resource_group_name         = azurerm_resource_group.this[each.key].name
  network_security_group_name = azurerm_network_security_group.this[each.key].name
}

resource "azurerm_network_security_rule" "deny_all_outbound" {
  for_each                    = var.instances
  name                        = "deny_all_outbound"
  priority                    = 4096
  direction                   = "Outbound"
  access                      = "Deny"
  protocol                    = "*"
  source_port_range           = "*"
  destination_port_range      = "*"
  source_address_prefix       = "*"
  destination_address_prefix  = "*"
  resource_group_name         = azurerm_resource_group.this[each.key].name
  network_security_group_name = azurerm_network_security_group.this[each.key].name
}

resource "azurerm_virtual_network" "this" {
  for_each            = var.instances
  name                = format("example-%s-vnet", each.key)
  resource_group_name = azurerm_resource_group.this[each.key].name
  location            = azurerm_resource_group.this[each.key].location
  tags                = var.tags
  address_space       = [each.value.address_space]
}

resource "azurerm_subnet" "this" {
  for_each             = var.instances
  name                 = "sqlmi"
  resource_group_name  = azurerm_resource_group.this[each.key].name
  virtual_network_name = azurerm_virtual_network.this[each.key].name
  address_prefixes     = [each.value.address_prefix]

  delegation {
    name = "sqlmi-delegation"

    service_delegation {
      name = "Microsoft.Sql/managedInstances"
      actions = [
        "Microsoft.Network/virtualNetworks/subnets/join/action",
        "Microsoft.Network/virtualNetworks/subnets/prepareNetworkPolicies/action",
        "Microsoft.Network/virtualNetworks/subnets/unprepareNetworkPolicies/action"
      ]
    }
  }
}

resource "azurerm_subnet_network_security_group_association" "this" {
  for_each                  = var.instances
  subnet_id                 = azurerm_subnet.this[each.key].id
  network_security_group_id = azurerm_network_security_group.this[each.key].id
}

resource "azurerm_route_table" "this" {
  depends_on = [
    azurerm_subnet.this
  ]

  for_each                      = var.instances
  name                          = format("example-%s-rt", each.key)
  resource_group_name           = azurerm_resource_group.this[each.key].name
  location                      = azurerm_resource_group.this[each.key].location
  tags                          = var.tags
  disable_bgp_route_propagation = false
}

resource "azurerm_subnet_route_table_association" "this" {
  for_each       = var.instances
  subnet_id      = azurerm_subnet.this[each.key].id
  route_table_id = azurerm_route_table.this[each.key].id
}

resource "azurerm_mssql_managed_instance" "primary" {
  depends_on = [
    azurerm_subnet_network_security_group_association.this,
    azurerm_subnet_route_table_association.this,
    azurerm_route_table.this
  ]

  name                = "example-primary-sqlmi"
  resource_group_name = azurerm_resource_group.this["primary"].name
  location            = azurerm_resource_group.this["primary"].location
  tags                = var.tags

  administrator_login          = "mradministrator"
  administrator_login_password = "thisIsDog11"

  license_type       = "BasePrice"
  sku_name           = "GP_Gen5"
  storage_size_in_gb = 32
  subnet_id          = azurerm_subnet.this["primary"].id
  vcores             = 4
}

resource "azurerm_mssql_managed_instance" "secondary" {
  depends_on = [
    azurerm_subnet_network_security_group_association.this,
    azurerm_subnet_route_table_association.this,
    azurerm_route_table.this
  ]

  name                = "example-secondary-sqlmi"
  resource_group_name = azurerm_resource_group.this["secondary"].name
  location            = azurerm_resource_group.this["secondary"].location
  tags                = var.tags

  administrator_login          = "mradministrator"
  administrator_login_password = "thisIsDog11"

  dns_zone_partner_id = azurerm_mssql_managed_instance.primary.id
  license_type        = "BasePrice"
  sku_name            = "GP_Gen5"
  storage_size_in_gb  = 32
  subnet_id           = azurerm_subnet.this["secondary"].id
  vcores              = 4
}

resource "azurerm_mssql_managed_instance_failover_group" "this" {
  name                        = "example-fog"
  location                    = azurerm_resource_group.this["secondary"].location
  managed_instance_id         = azurerm_mssql_managed_instance.primary.id
  partner_managed_instance_id = azurerm_mssql_managed_instance.secondary.id

  read_write_endpoint_failover_policy {
    grace_minutes = 60
    mode          = "Automatic"
  }
}
```

### Panic Output

N/A

### Expected Behaviour

Failover group should have been successfully created.

### Actual Behaviour

`apply` failed when attempting to create the failover group:

```
│ Error: creating Instance Failover Group: (Name "example-fog" / Location Name "centralus" / Resource Group "example-primary-rg"): sql.InstanceFailoverGroupsClient#CreateOrUpdate: Failure sending request: StatusCode=500 -- Original Error: Code="InternalServerError" Message="An unexpected error occured while processing the request. Tracking ID: 'xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx'"
│ 
│   with azurerm_mssql_managed_instance_failover_group.this,
│   on main.tf line 261, in resource "azurerm_mssql_managed_instance_failover_group" "this":
│  261: resource "azurerm_mssql_managed_instance_failover_group" "this" {
│ 
│ creating Instance Failover Group: (Name "example-fog" / Location Name "centralus" / Resource Group "example-primary-rg"): sql.InstanceFailoverGroupsClient#CreateOrUpdate: Failure sending request:
│ StatusCode=500 -- Original Error: Code="InternalServerError" Message="An unexpected error occured while processing the request. Tracking ID: 'xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx'"
```

A subsequent `plan` run indicates that the secondary instance needs to be recreated, which leads me to believe that there may be an issue related with `dns_zone_partner_id` when it is initially created. I also tried manually updating the state file, which resulted in the same plan. 

```
Terraform will perform the following actions:

  # azurerm_mssql_managed_instance.secondary must be replaced
-/+ resource "azurerm_mssql_managed_instance" "secondary" {
      + dns_zone_partner_id          = "/subscriptions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/resourceGroups/example-primary-rg/providers/Microsoft.Sql/managedInstances/example-primary-sqlmi" # forces replacement
      ~ fqdn                         = "example-secondary-sqlmi.xxxxxxxxxxxx.database.windows.net" -> (known after apply)
      ~ id                           = "/subscriptions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/resourceGroups/example-secondary-rg/providers/Microsoft.Sql/managedInstances/example-secondary-sqlmi" -> (known after apply)
        name                         = "example-secondary-sqlmi"
        tags                         = {
        }
        # (15 unchanged attributes hidden)
    }

  # azurerm_mssql_managed_instance_failover_group.this will be created
  + resource "azurerm_mssql_managed_instance_failover_group" "this" {
      + id                                        = (known after apply)
      + location                                  = "centralus"
      + managed_instance_id                       = "/subscriptions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/resourceGroups/example-primary-rg/providers/Microsoft.Sql/managedInstances/example-primary-sqlmi"
      + name                                      = "example-fog"
      + partner_managed_instance_id               = (known after apply)
      + partner_region                            = (known after apply)
      + readonly_endpoint_failover_policy_enabled = true
      + role                                      = (known after apply)

      + read_write_endpoint_failover_policy {
          + grace_minutes = 60
          + mode          = "Automatic"
        }
    }

Plan: 2 to add, 0 to change, 1 to destroy.
```

### Steps to Reproduce

1. `terraform apply`

### Important Factoids

The initial plan reflects the `dns_zone_partner_id` value for the secondary instance is being set:

```hcl
  # azurerm_mssql_managed_instance.primary will be created
  + resource "azurerm_mssql_managed_instance" "primary" {
      + administrator_login          = "mradministrator"
      + administrator_login_password = (sensitive value)
      + collation                    = "SQL_Latin1_General_CP1_CI_AS"
      + fqdn                         = (known after apply)
      + id                           = (known after apply)
      + license_type                 = "BasePrice"
      + location                     = "eastus2"
      + minimum_tls_version          = "1.2"
      + name                         = "example-primary-sqlmi"
      + proxy_override               = "Default"
      + public_data_endpoint_enabled = false
      + resource_group_name          = "example-primary-rg"
      + sku_name                     = "GP_Gen5"
      + storage_account_type         = "GRS"
      + storage_size_in_gb           = 32
      + subnet_id                    = "/subscriptions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/resourceGroups/example-primary-rg/providers/Microsoft.Network/virtualNetworks/example-primary-vnet/subnets/sqlmi"
      + tags                         = {
        }
      + timezone_id                  = "UTC"
      + vcores                       = 4
    }

  # azurerm_mssql_managed_instance.secondary will be created
  + resource "azurerm_mssql_managed_instance" "secondary" {
      + administrator_login          = "mradministrator"
      + administrator_login_password = (sensitive value)
      + collation                    = "SQL_Latin1_General_CP1_CI_AS"
      + dns_zone_partner_id          = (known after apply)
      + fqdn                         = (known after apply)
      + id                           = (known after apply)
      + license_type                 = "BasePrice"
      + location                     = "centralus"
      + minimum_tls_version          = "1.2"
      + name                         = "example-secondary-sqlmi"
      + proxy_override               = "Default"
      + public_data_endpoint_enabled = false
      + resource_group_name          = "example-secondary-rg"
      + sku_name                     = "GP_Gen5"
      + storage_account_type         = "GRS"
      + storage_size_in_gb           = 32
      + subnet_id                    = "/subscriptions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/resourceGroups/example-secondary-rg/providers/Microsoft.Network/virtualNetworks/example-secondary-vnet/subnets/sqlmi"
      + tags                         = {
        }
      + timezone_id                  = "UTC"
      + vcores                       = 4
    }

  # azurerm_mssql_managed_instance_failover_group.this will be created
  + resource "azurerm_mssql_managed_instance_failover_group" "this" {
      + id                                        = (known after apply)
      + location                                  = "centralus"
      + managed_instance_id                       = (known after apply)
      + name                                      = "example-fog"
      + partner_managed_instance_id               = (known after apply)
      + partner_region                            = (known after apply)
      + readonly_endpoint_failover_policy_enabled = true
      + role                                      = (known after apply)

      + read_write_endpoint_failover_policy {
          + grace_minutes = 60
          + mode          = "Automatic"
        }
    }
```

### References

* #15203

**Comment by fcatacut** [comment]
@aristosvo, are you able to provide any insight since you implemented the `azurerm_sql_managed_instance_failover_group` resource?

**Comment by aristosvo** [comment]
@fcatacut Not really, I'm afraid.

The tests for the resource were working fine at the time, and can be found [here](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/internal/services/sql/sql_managed_instance_failover_group_resource_test.go).

**Comment by fgarcia-cnb** [comment]
the azurerm_sql_managed_instance_failover_group resource appears to work, but i think these resources are going to be deprecated soon and replaced by the azurerm_sql_managed_instance_failover_group

**Comment by anhlqn** [comment]
I'm running into exactly the same issue with the `azurerm_mssql_managed_instance` resource. Haven't tried the old `azurerm_sql_managed_instance` yet, but is it possible to migrate to the new `azurerm_mssql_managed_instance` later? I tested with `terraform state mv` but it failed with an error that they were two different resource types so the terraform state couldn't be renamed.

**Comment by fcatacut** [comment]
I can confirm that `azurerm_sql_managed_instance` worked as of AzureRM 2.96. We were about to release a module to our product teams that referenced that resource when `azurerm_mssql_managed_instance` was released. Thinking that it was just a simple resource change, we held off releasing because we wanted to eliminate the need to run `terraform state mv` commands if at all possible.

I think that provisioning just the secondary/failover server using `azurerm_sql_managed_instance` will resolve the issue, but I haven't had a chance to verify this since it's an 8+ hour deployment.

**Comment by harkau779** [comment]
I also faced the exact same issue. 
I deployed primary & secondary managed instance using `azurerm_mssql_managed_instance` & failover group using `azurerm_mssql_managed_instance_failover_group` but failover_group deployment failed with above mentioned error. Then I tried deploying failover group with `azurerm_sql_managed_instance_failover_group` using same AzureRM 2.99.0 provider & it worked. So, it seems that the issue is with `azurerm_mssql_managed_instance_failover_group` resource block.

**Comment by harkau779** [comment]
> I can confirm that `azurerm_sql_managed_instance` worked as of AzureRM 2.96. We were about to release a module to our product teams that referenced that resource when `azurerm_mssql_managed_instance` was released. Thinking that it was just a simple resource change, we held off releasing because we wanted to eliminate the need to run `terraform state mv` commands if at all possible.
> 
> I think that provisioning just the secondary/failover server using `azurerm_sql_managed_instance` will resolve the issue, but I haven't had a chance to verify this since it's an 8+ hour deployment.

I tried provisioning just the secondary/failover server using `azurerm_sql_managed_instance`  but still facing the same issue.

**Comment by macro-labs** [comment]
I am seeing the same issue with exactly the same error. 

I created both SQL Managed Instances using azurerm_mssql_managed_instance. They are each in different, paired regions. When creating the secondary instance, I properly set the dns_zone_partner_id (Failing to do this is a common cause of failover group creation errors in Azure) and manually confirmed that it worked by examining the fqdns of both the primary and secondary instances. As expected, both instances share the same SQL Managed Instance subdomain - which is what the practical effect of setting dns_zone_partner_id really is. The networks are paired with global vNet peering, and there are are NSGs connected to each hosting subnet that include a rules that open ports 1433, 5022 and 11000-11999 inbound and outbound to enable replication and redirection. Net: the setup passes a pretty exhaustive inspection for readiness for failover group creation.

I am able to create the failover group in the Azure portal with no problems. Azure takes fewer that 3 minutes to complete setting it up and begin synchronizing, so there is no reason why the provider should error out on this. When using azurerm_mssql_managed_instance_failover_group (azurerm provider v3.02) I get the error noted above - and only after twice that long.

**Comment by fcatacut** [comment]
@katbyte I believe that this issue was resolved by PR #16705, which was released as part of v3.6.0.

**Comment by tombuildsstuff** [comment]
Closing per @fcatacut's comment above - thanks for letting us know
